init.sh: move netlink device creation to Go

rgo3 opened the pull request:
vendor: update vishvananda/netlink:
```
This pulls in the following recent changes:

- vishvananda/netlink@88c0de
- vishvananda/netlink@16d31d
- vishvananda/netlink@6afddb3

These are needed to move ipip and tunnel device creation from
init.sh to Go.

Signed-off-by: Robin Gögge <r.goegge@isovalent.com>
```
init.sh: move ipip device creation to Go:
```
This refactors the creation of the cilium_ipip{4,6} devices to a
Go based implementation and adds a simple device creation test.

One behavioural change was made: no longer try to repurpose
tunl0 by renaming it and enabling collect_md, as that dance is
rather tricky and fragile due to the kernel module needing to be
loaded first.

With this change, we still prefix the fallback devices with cilium_,
meaning cilium_tunl0 for v4 and cilium_ip6tnl0 for v6, but we no
longer try to use or reconfigure these devices. cilium_ipip4/6 are
always created with the correct parameters from the start, reducing
the amount of moving parts.

Signed-off-by: Robin Gögge <r.goegge@isovalent.com>
Co-authored-by: Timo Beckers <timo@isovalent.com>
```

init.sh: move tunnel device creation to Go

```
This refactors the creation of the cilium_{vxlan,geneve} devices to a
Go-based implementation and adds a simple device creation test.

Signed-off-by: Robin Gögge <r.goegge@isovalent.com>
```
Fixes: #26254 #20740

rgo3 committed:
init.sh: move ipip device creation to Go

This refactors the creation of the cilium_ipip{4,6} devices to a
Go-based implementation and adds a simple device creation test.

One behavioural change was made: no longer try to repurpose
tunl0 by renaming it and enabling collect_md, as that dance is
rather tricky and fragile due to the kernel module needing to be
loaded first.

With this change, we still prefix the fallback devices with cilium_,
meaning cilium_tunl0 for v4 and cilium_ip6tnl0 for v6, but we no
longer try to use or reconfigure these devices. cilium_ipip4/6 are
always created with the correct parameters from the start, reducing
the amount of moving parts.

Signed-off-by: Robin Gögge <r.goegge@isovalent.com>
Co-authored-by: Timo Beckers <timo@isovalent.com>

rgo3 committed:
vendor: update vishvananda/netlink

This pulls in the following recent changes:

- vishvananda/netlink@88c0defd466a5d755d6c3956617ae27217bdda87
- vishvananda/netlink@16d31db2358884ece09bbbaa2b754c225c788032
- vishvananda/netlink@6afddb37c1f00693528264222a57a3c08dccbab7

These are needed to move ipip and tunnel device creation from
init.sh to Go.

Signed-off-by: Robin Gögge <r.goegge@isovalent.com>

rgo3 committed:
init.sh: move tunnel device creation to Go

This refactors the creation of the cilium_{vxlan,geneve} devices to a
Go-based implementation and adds a simple device creation test.

Signed-off-by: Robin Gögge <r.goegge@isovalent.com>

ti-mo commented on pkg/datapath/linux/config/config.go:
```diff
@@ -153,6 +153,10 @@ func (h *HeaderfileWriter) WriteNodeConfig(w io.Writer, cfg *datapath.LocalNodeC
 	cDefinesMap["TUNNEL_PROTOCOL"] = fmt.Sprintf("%d", tunnelProtocols[encapProto])
 	cDefinesMap["TUNNEL_PORT"] = fmt.Sprintf("%d", option.Config.TunnelPort)
 
+	if tunnelDev, err := netlink.LinkByName(fmt.Sprintf("cilium_%s", encapProto)); err == nil {
```
I don't think this should be optional. We should probably gate this on `option.Config.TunnelingEnabled()` and bail out on any errors.

rgo3 replied:
I agree on this not being optimal but I was unsure how to deal with this properly given that `TunnelingEnabled()` can return false while `encapProto` can still be set to `geneve`.  That's why I am gating the define statement being added on the existence of either one of the necessary devices.

ti-mo commented on pkg/datapath/loader/base.go:
```diff
@@ -361,6 +361,23 @@ func (l *Loader) Reinitialize(ctx context.Context, o datapath.BaseProgramOwner,
+		}
+	}
+	if !option.Config.TunnelingEnabled() &&
+		option.Config.EnableNodePort &&
```
Looks like this could this be folded into the previous if block. Maybe make this a switch instead, since only one encapProto can be selected?

rgo3 replied:
I just moved this code up a few lines and Julian is cleaning up this section anyway: https://github.com/cilium/cilium/pull/26898.

julianwiedmann replied:
👋 would it help if I clean that PR up and get it in first?

ti-mo replied:
No worries, it can be done later. This code just looks like it needs some love. :wink:

ti-mo commented on pkg/datapath/loader/netlink.go:
```diff
@@ -447,6 +447,131 @@ func addHostDeviceAddr(hostDev netlink.Link, ipv4, ipv6 net.IP) error {
+
+// setupGeneveDevice ensures the cilium_geneve device is created with the correct
+// port and mtu values.
+func setupGeneveDevice(port, mtu int) error {
```
This shares quite a bit of code with `setupVxlanDevice()`, could this be made a little shorter?

rgo3 replied:
I've factored out similarities into a separate function now.

ti-mo commented on pkg/datapath/loader/netlink_test.go:
```diff
@@ -114,6 +114,139 @@ func TestSetupDev(t *testing.T) {
+			link, err := netlink.LinkByName(defaults.GeneveDevice)
+			require.NoError(t, err)
+
+			geneve, isGeneve := link.(*netlink.Geneve)
```
Nit: `geneve, ok := `... is a more common pattern.

ti-mo reviewed (approved):
Few nits to address while waiting for LVH builds. :grin: Nice work!

ti-mo committed:
route: convert TestListRules to testing.T

This was failing after the netlink lib upgrade and wasn't giving any output
whatsoever.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
route: assign family in TestListRules

As of vishvananda/netlink@dbf1bd04f9af9c9a96501ab340cb90d928c9e6a7
("Linux rule list results have Family populated"), rule lists have the
Family field populated, making TestListRules fail on comparing the result.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo reviewed (approved):
This is in good enough shape, let's get it merged!

lmb commented on pkg/datapath/loader/base.go:
```diff
@@ -361,6 +361,23 @@ func (l *Loader) Reinitialize(ctx context.Context, o datapath.BaseProgramOwner,
 	}
 
+	if !option.Config.TunnelingEnabled() {
+		if option.Config.EnableIPv4EgressGateway || option.Config.EnableHighScaleIPcache {
```
Please use `option.Config.EgressGatewayCommonEnabled()`

ti-mo replied:
Thanks! `option.Config.EnableIPv4EgressGateway` is read in a bunch of places throughout the codebase, though. Better patch it up if that's not supposed to happen, or folks will copy-paste.

ti-mo replied:
Ah, looks like you changed this line in particular last week, and it probably got dropped in conflict resolution during rebase.

thorn3r reviewed (approved):
nice work :+1: death to init.sh :skull:
